### PR TITLE
BAEL-7678 Find max value in Spring Data JPA

### PR DESCRIPTION
**Pull request by Saber-k**

### CODE REVIEW

File: persistence-modules/spring-data-jpa-query-4/src/test/java/com/baeldung/spring/data/jpa/maxvalue/FindingMaxValueIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    private EmployeeMaxValueService employeeMaxValueService;
+
+    @Autowired
+    private EmployeeRepository employeeRepository;
+
+    private final static Long EXPECTED_VALUE = 800L;
+
+    @BeforeEach
+    void givenEmployees() {
+        // given
```
Comment (theangrydev) [inline persistence-modules/spring-data-jpa-query-4/src/test/java/com/baeldung/spring/data/jpa/maxvalue/FindingMaxValueIntegrationTest.java:31]:
```suggestion
```

File: persistence-modules/spring-data-jpa-query-4/src/test/java/com/baeldung/spring/data/jpa/maxvalue/FindingMaxValueIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        Employee employee3 = new Employee(3, "Jack", 600L);
+
+        employeeRepository.saveAll(List.of(employee1, employee2, employee3));
+    }
+
+    @Test
+    void givenEmployees_whenFindMaxUsingDerivedQuery_thenMaxReturned() {
+        Optional<Employee> result = employeeRepository.findTopByOrderBySalaryDesc();
+
+        Assertions.assertThat(result)
```
Comment (theangrydev) [inline persistence-modules/spring-data-jpa-query-4/src/test/java/com/baeldung/spring/data/jpa/maxvalue/FindingMaxValueIntegrationTest.java:43]: Add the static imports for Assertions. (same elsewhere)
```suggestion
        assertThat(result)
```

File: persistence-modules/spring-data-jpa-query-4/src/main/java/com/baeldung/spring/data/jpa/maxvalue/EmployeeMaxValueService.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Root;
+import java.util.Optional;
+
+@Service
+public class EmployeeMaxValueService {
+    private final EntityManager entityManager;
+
+    public EmployeeMaxValueService(EntityManager entityManager) {
```
Comment (theangrydev) [inline persistence-modules/spring-data-jpa-query-4/src/main/java/com/baeldung/spring/data/jpa/maxvalue/EmployeeMaxValueService.java:16]: Should we have `@Autowired` on the constructor for clarity about where the dependency comes from?

```suggestion
    @Autowired
    public EmployeeMaxValueService(EntityManager entityManager) {
```

File: persistence-modules/spring-data-jpa-query-4/src/main/java/com/baeldung/spring/data/jpa/maxvalue/EmployeeRepository.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.Comparator;
+import java.util.Optional;
+
+public interface EmployeeRepository extends JpaRepository<Employee, Integer> {
+    Optional<Employee> findTopByOrderBySalaryDesc();
+
+    Optional<EmployeeSalary> findTopSalaryByOrderBySalaryDesc();
+
+    @Query("SELECT MAX(e.salary) FROM Employee e")
+    Optional<Long> findTopSalaryJQPL();
```
Comment (theangrydev) [inline persistence-modules/spring-data-jpa-query-4/src/main/java/com/baeldung/spring/data/jpa/maxvalue/EmployeeRepository.java:17]: We follow https://google.github.io/styleguide/javaguide.html#s5.3-camel-case for naming conventions and so this should be Jqpl not JQPL

File: persistence-modules/spring-data-jpa-query-4/src/main/java/com/baeldung/spring/data/jpa/maxvalue/Employee.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Employee employee = (Employee) o;
+        return Objects.equals(id, employee.id) && Objects.equals(name, employee.name) && Objects.equals(salary, employee.salary);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, name, salary);
+    }
```
Comment (theangrydev) [inline persistence-modules/spring-data-jpa-query-4/src/main/java/com/baeldung/spring/data/jpa/maxvalue/Employee.java:60]: Are these necessary for the production code to work or just used for the test assertions?

**Review by theangrydev — CHANGES_REQUESTED** [review]

File: persistence-modules/spring-data-jpa-query-4/src/main/java/com/baeldung/spring/data/jpa/maxvalue/Employee.java (hunk 5 above)
Comment (Saber-k) [reply persistence-modules/spring-data-jpa-query-4/src/main/java/com/baeldung/spring/data/jpa/maxvalue/Employee.java]: not necessary, just auto-generated via IJ, removed

**Review by theangrydev — APPROVED** [review]

**Comment by theangrydev** [comment]
retest this

**Pull request merged**